openwisp_monitoring/device/apps.py

```diff
@@ -30,3 +51,26 @@ def device_post_delete_receiver(cls, instance, **kwargs):
         instance.__class__ = DeviceData
         instance.checks.all().delete()
         instance.metrics.all().delete()
+
+    @classmethod
+    def _generate_cache_key(cls, device):
+        return f'device-{device.pk}-react-to-updates'
+
+    @classmethod
+    def manage_device_recovery_cache_key(cls, instance, status, **kwargs):
+        """
+        WIP
+        """
+        cache_key = cls._generate_cache_key(device=instance.device)
+        if status == 'critical':
+            if app_settings.DEVICE_RECOVERY_DETECTION:
```

Comment:
maybe this is not necessary here?

Reply:
The line is not required in the module as such but was being required while unit testing. The reason being the setting `DEVICE_RECOVERY_DETECTION` is `True` by default and hence the function `manage_device_recovery_cache_key` gets connected to `health_status_changed` signal when the app is loaded right before unit tests start. I was thinking of some better way to avoid the above and finally did it as below:
1. Disconnect the signal in the test.
2. Reload the app to verify that function doesn't get connected to signal when setting is patched as `False`.
3. After verifying that cache key has not been set, reload the app with setting as `True` (Default),

You may check `test_device_recovery_cache_key_not_set` in `test_recovery.TestRecovery` for the code. It might seem a bit lengthy but checks the condition in a better manner than above workaround :smiley: